crates/mako/src/plugins/duplicate_package_checker.rs

```diff
@@ -0,0 +1,189 @@
+    let json: Value = serde_json::from_str(&content)?;
+
+    let name = json["name"]
+        .as_str()
+        .ok_or("Missing package name")?
+        .to_string();
+    let version = json["version"].as_str().ok_or("Missing package version")?;
+    let version = Version::parse(version)?;
+
+    Ok(PackageInfo {
+        name,
+        version,
+        path: path.parent().unwrap().to_path_buf(),
+    })
+}
+
+/// Cleans the path by replacing /node_modules/ or \node_modules\ with /~/
+fn clean_path(module_path: &Path) -> PathBuf {
+    let path_str = module_path.to_string_lossy();
+    let cleaned_path = path_str
+        .replace("/node_modules/", "/~/")
+        .replace("\\node_modules\\", "/~/");
+    PathBuf::from(cleaned_path)
+}
+
+/// Makes the cleaned path relative to the given context
+fn clean_path_relative_to_context(module_path: &Path, context: &Path) -> PathBuf {
+    let cleaned_path = clean_path(module_path);
+    let context_str = context.to_str().unwrap();
+    let cleaned_path_str = cleaned_path.to_str().unwrap();
+
+    if cleaned_path_str.starts_with(context_str) {
+        let relative_path = cleaned_path_str.trim_start_matches(context_str);
+        PathBuf::from(format!(".{}", relative_path))
+    } else {
+        cleaned_path
+    }
+}
+
+impl DuplicatePackageCheckerPlugin {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn verbose(mut self, verbose: bool) -> Self {
+        self.verbose = verbose;
+        self
+    }
+
+    pub fn show_help(mut self, show_help: bool) -> Self {
+        self.show_help = show_help;
+        self
+    }
+
+    pub fn emit_error(mut self, emit_error: bool) -> Self {
+        self.emit_error = emit_error;
+        self
+    }
+
+    // find closest package.json file
+    fn find_closet_package(entry: &Path) -> Result<PackageInfo, Box<dyn Error>> {
+        let mut current = entry;
+        loop {
+            let package_json = current.join("package.json");
+            if package_json.exists() {
+                return read_package_info(&package_json);
+            }
+
+            current = current.parent().ok_or("No package.json found")?;
+        }
+    }
+
+    fn find_duplicates(packages: Vec<PackageInfo>) -> HashMap<String, Vec<PackageInfo>> {
+        let mut package_map: HashMap<String, Vec<PackageInfo>> = HashMap::new();
+
+        for package in packages {
+            package_map
+                .entry(package.name.clone())
+                .or_default()
+                .push(package);
+        }
+
+        package_map
+            .into_iter()
+            .filter(|(_, versions)| versions.len() > 1)
+            .collect()
+    }
+
+    fn check_duplicates(
+        &self,
+        module_graph: &RwLock<ModuleGraph>,
+    ) -> HashMap<String, Vec<PackageInfo>> {
+        let mut packages = Vec::new();
+
+        for module in module_graph.read().unwrap().modules() {
+            let module_info = module.info.as_ref().unwrap();
```

Comment:
use `moduleInfo#resolved_resource`  is much more easier

Reply:
> use `moduleInfo#resolved_resource` is much more easier

+1

Reply:
@stormslowly @xusd320 I am confused with this. How does it work?

Reply:
@jeasonnow moduleInfo#resolved_resource does contain the package.json info, we don't need to resolve the package.json manually.

Reply:
done. @sorrycc @stormslowly